src/passes/OptimizeInstructions.cpp

```diff
@@ -1345,15 +1381,49 @@ struct OptimizeInstructions
         }
       }
       // operations on all 1s
-      // TODO: shortcut method to create an all-ones?
       if (right->value == Literal(int32_t(-1)) ||
           right->value == Literal(int64_t(-1))) {
         if (binary->op == Abstract::getBinary(type, Abstract::And)) {
+          // x & -1   ==>   x
           return binary->left;
         } else if (binary->op == Abstract::getBinary(type, Abstract::Or) &&
                    !EffectAnalyzer(getPassOptions(), features, binary->left)
                       .hasSideEffects()) {
+          // x | -1   ==>   -1
           return binary->right;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::Sub)) {
+          // x - (-1)   ==>   x + 1
+          binary->op = Abstract::getBinary(type, Abstract::Add);
+          right->value = Literal::makeFromInt32(1, type);
+          return binary;
+        } else if ((binary->op == Abstract::getBinary(type, Abstract::RemS) ||
+                    binary->op == Abstract::getBinary(type, Abstract::GtU)) &&
+                   !EffectAnalyzer(getPassOptions(), features, binary->left)
+                      .hasSideEffects()) {
+          // (signed)x % -1     ==>   0
+          // (unsigned)x > -1   ==>   0
+          right->value = Literal::makeSingleZero(type);
+          return right;
+        } else if (binary->op == Abstract::getBinary(type, Abstract::LtU)) {
+          // (unsigned)x < -1   ==>   x != -1
+          binary->op = Abstract::getBinary(type, Abstract::Ne);
+          return binary;
```

Comment:
please explain why this helps - it doesn't change the size, so it's not obvious to me what the benefit is?

Reply:
1. Sync with LLVM rules
2. Improve asm.js size: https://github.com/WebAssembly/binaryen/pull/2870/files#diff-557cae6822e26d5d32c9ed56e42459fbR43

Reply:
I see, thanks. Worth mentioning these in a comment (for asm.js I would say `friendlier to JS emitting as we don't need to write an unsigned -1 value which is large`).

Reply:
Sure. Will do